core/src/main/scala/org/apache/spark/deploy/Client.scala

```diff
@@ -17,33 +17,40 @@
 
 package org.apache.spark.deploy
 
-import scala.concurrent._
+import scala.concurrent.ExecutionContext
+import scala.reflect.ClassTag
+import scala.util.{Failure, Success}
 
-import akka.actor._
-import akka.pattern.ask
-import akka.remote.{AssociationErrorEvent, DisassociatedEvent, RemotingLifecycleEvent}
 import org.apache.log4j.{Level, Logger}
 
+import org.apache.spark.rpc.{RpcEndpointRef, RpcAddress, RpcEnv, ThreadSafeRpcEndpoint}
 import org.apache.spark.{Logging, SecurityManager, SparkConf}
 import org.apache.spark.deploy.DeployMessages._
 import org.apache.spark.deploy.master.{DriverState, Master}
-import org.apache.spark.util.{ActorLogReceive, AkkaUtils, Utils}
+import org.apache.spark.util.{SparkExitCode, Utils}
 
 /**
  * Proxy that relays messages to the driver.
  */
-private class ClientActor(driverArgs: ClientArguments, conf: SparkConf)
-  extends Actor with ActorLogReceive with Logging {
-
-  var masterActor: ActorSelection = _
-  val timeout = AkkaUtils.askTimeout(conf)
-
-  override def preStart(): Unit = {
-    masterActor = context.actorSelection(
-      Master.toAkkaUrl(driverArgs.master, AkkaUtils.protocol(context.system)))
-
-    context.system.eventStream.subscribe(self, classOf[RemotingLifecycleEvent])
-
+private class ClientEndpoint(
+    override val rpcEnv: RpcEnv,
+    driverArgs: ClientArguments,
+    masterEndpoint: RpcEndpointRef,
+    conf: SparkConf)
+  extends ThreadSafeRpcEndpoint with Logging {
+
+  private val forwardMessageThread =
+    Utils.newDaemonSingleThreadScheduledExecutor("client-forward-message")
+  private implicit val forwardMessageExecutionContext =
```

Comment:
who uses this? can we make it non-implicit instead and explicitly call this variable when we use it? Being implicit makes it fairly hard to review/change in the future.